pkg/services/gc/gc.go

```diff
@@ -0,0 +1,389 @@
+package gc
+
+import (
+	"context"
+	"fmt"
+	"slices"
+	"strings"
+	"sync"
+
+	"github.com/go-logr/logr"
+	"golang.org/x/exp/maps"
+	authorizationv1 "k8s.io/api/authorization/v1"
+	k8serr "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	ctrlCli "sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/opendatahub-io/opendatahub-operator/v2/pkg/controller/client"
+)
+
+// TODO: this singleton must be removed.
+var Instance *GC
```

Comment:
Can this eventually be a struct?

Reply:
it is actually a struct, I've removed the global `Run` method and added some better comments